src/gui/systray.cpp

```diff
@@ -133,123 +126,225 @@ void Systray::setToolTip(const QString &tip)
+    const auto distLeft = trayIconCenter.x() - screenRect.left();
+    const auto distTop = trayIconCenter.y() - screenRect.top();
 
-    if (availableHeight < screenHeight) {
-        // taskbar is on top or bottom
-        if (trayIconTopCenterX + (400 * 0.5) > availableWidth) {
-            return availableWidth - 400 - 12;
-        } else {
-            return trayIconTopCenterX - (400 * 0.5);
-        }
+    const auto minDist = std::min({distRight, distTop, distBottom});
+
+    if (minDist == distBottom) {
+        return TaskBarPosition::Bottom;
+    } else if (minDist == distLeft) {
+        return TaskBarPosition::Left;
+    } else if (minDist == distTop) {
+        return TaskBarPosition::Top;
     } else {
-        if (trayScreen->availableGeometry().x() > trayScreen->geometry().x()) {
-            // on the left
-            return (screenWidth - availableWidth) + 6;
-        } else {
-            // on the right
-            return screenWidth - 400 - (screenWidth - availableWidth) - 6;
-        }
+        return TaskBarPosition::Right;
     }
 #endif
 }
-int Systray::calcTrayWindowY()
+
+// TODO: Get real taskbar dimensions Linux as well
+QRect Systray::taskbarGeometry() const
 {
-#ifdef Q_OS_OSX
-    // macOS menu bar is always 22 (effective) pixels
-    // don't use availableGeometry() here, because this also excludes the dock
-    return 22+6;
+#if defined(Q_OS_WIN)
+    QRect tbRect = Utility::getTaskbarDimensions();
+    //QML side expects effective pixels, convert taskbar dimensions if necessary
+    auto pixelRatio = currentScreen()->devicePixelRatio();
+    if (pixelRatio != 1) {
+        tbRect.setHeight(tbRect.height() / pixelRatio);
+        tbRect.setWidth(tbRect.width() / pixelRatio);
+    }
+    return tbRect;
+#elif defined(Q_OS_MACOS)
+    // Finder bar is always 22px height on macOS (when treating as effective pixels)
+    auto screenWidth = currentScreenRect().width();
+    return QRect(0, 0, screenWidth, 22);
 #else
-    QScreen* trayScreen = nullptr;
-    if (QGuiApplication::screens().count() > 1) {
-        trayScreen = QGuiApplication::screens().at(screenIndex());
+    if (taskbarOrientation() == TaskBarPosition::Bottom || taskbarOrientation() == TaskBarPosition::Top) {
+        auto screenWidth = currentScreenRect().width();
+        return QRect(0, 0, screenWidth, 32);
     } else {
-        trayScreen = QGuiApplication::primaryScreen();
+        auto screenHeight = currentScreenRect().height();
+        return QRect(0, 0, 32, screenHeight);
     }
+#endif
+}
 
-    int screenHeight = trayScreen->geometry().height();
-    int availableHeight = trayScreen->availableGeometry().height();
+QRect Systray::currentScreenRect() const
+{
+    const auto screen = currentScreen();
+    const auto rect = screen->geometry();
+#if defined(Q_OS_WIN) || defined(Q_OS_MACOS)
+    return rect;
+#else
+    return rect.translated(screen->virtualGeometry().topLeft());
```

Comment:
I don't get why Windows/MacOS is treated differently than all others ("probably Linux" :smile:). A short comment in the code would help. If the complicated `rect.translated(screen->virtualGeometry().topLeft()) == rect` on Windows/MacOS, I would always return `rect.translated(screen->virtualGeometry().topLeft())` and avoid preprocessor directives. In any case, I would want a short comment why this translation is necessary.

Reply:
Good point, I didn't think of checking but this is always == rect on Windows (and I assume on MacOS as well, it's very likely not that different).

Reply:
OK, turn out I was wrong, this was maybe necessary: see #2166 

On Windows at least if the task bar is moved on a different screen it works fine (this is what I tested), but it doesn't if the primary screen is changed. Investigating further but I might have to reintroduce that hunk.